lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyHttpServerEventListener.java

```diff
@@ -0,0 +1,155 @@
+    final HttpHeaders headers;
+    final HttpRequest request;
+    if(msg instanceof HttpRequest) {
+      LOG.log(Level.FINEST, "HttpRequest received");
+      request = (HttpRequest) msg;
+      this.httpRequest = request;
+      headers = request.headers();
+      if (!headers.isEmpty()) {
+        for (Map.Entry<String, String> h: headers) {
+          CharSequence key = h.getKey();
+          CharSequence value = h.getValue();
+          buf.append("HEADER: ").append(key).append(" = ").append(value).append("\r\n");
+        }
+        buf.append("\r\n");
+      }
+      appendDecoderResult(buf, request);
+    } else if (msg instanceof HttpContent) {
+      LOG.log(Level.FINEST, "HttpContent received");
+      HttpContent httpContent = (HttpContent) msg;
+      ByteBuf content = httpContent.content();
+      if (content.isReadable()) {
+        buf.append("CONTENT: ");
+        buf.append(content.toString(CharsetUtil.UTF_8));
+        buf.append("\r\n");
+        appendDecoderResult(buf, this.httpRequest);
+      }
+
+      if (msg instanceof LastHttpContent) {
+        buf.append("END OF CONTENT\r\n");
+        LastHttpContent trailer = (LastHttpContent) msg;
+        if (!trailer.trailingHeaders().isEmpty()) {
+          buf.append("\r\n");
+          for (CharSequence name: trailer.trailingHeaders().names()) {
+            for (CharSequence value: trailer.trailingHeaders().getAll(name)) {
+              buf.append("TRAILING HEADER: ");
+              buf.append(name).append(" = ").append(value).append("\r\n");
+            }
+          }
+          buf.append("\r\n");
+        }
+
+        final Channel channel = ctx.channel();
+        byte[] message = new byte[content.readableBytes()];
+        content.readBytes(message);
+        if (LOG.isLoggable(Level.FINEST)) {
+          LOG.log(Level.FINEST, "MessageEvent: local: {0} remote: {1} :: {2}", new Object[]{
+                  channel.localAddress(), channel.remoteAddress(), content});
+        }
+
+        if (message.length > 0) {
+          // send to the dispatch stage
+          this.stage.onNext(this.getTransportEvent(message, channel));
+        }
+      }
+    } else {
+      LOG.log(Level.SEVERE, "Unknown type of message received: {0}", msg);
+    }
+    LOG.log(Level.FINEST, buf.toString());
+  }
+
+  private static void appendDecoderResult(final StringBuilder buf, final HttpObject o) {
+    DecoderResult result = o.getDecoderResult();
+    if (result.isSuccess()) {
+      return;
+    }
+
+    buf.append(".. WITH DECODER FAILURE: ");
+    buf.append(result.cause());
+    buf.append("\r\n");
+  }
+
+
+
+  @Override
+  protected TransportEvent getTransportEvent(final byte[] message, final Channel channel) {
+    return new TransportEvent(message, new NettyLink<>(channel, new ByteEncoder()));
+  }
+
+  @Override
+  protected void exceptionCleanup(final ChannelHandlerContext ctx, final Throwable cause) {
```

Comment:
It is required method to extend `AbstractNettyEventyListener` and it only performs close channel when it is clientListener.